pkg/api/validate/dynamic/serviceprincipal.go

```diff
@@ -8,11 +8,45 @@ import (
 	"net/http"
 
 	"github.com/form3tech-oss/jwt-go"
+	"github.com/sirupsen/logrus"
 
 	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
 	"github.com/Azure/ARO-RP/pkg/util/azureclaim"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
 )
 
+type ServicePrincipalValidatorInterface interface {
+	Validate(ctx context.Context, log *logrus.Entry, azEnv *azureclient.AROEnvironment, tokenClient aad.TokenClient, clientID, clientSecret, tenantID string) error
+}
+
+type SPValidator struct {
+}
+
+func (v SPValidator) Validate(ctx context.Context, log *logrus.Entry, azEnv *azureclient.AROEnvironment, tokenClient aad.TokenClient, clientID, clientSecret, tenantID string) error {
```

Comment:
Try to apply the same idea to other methods in the denymaic validator. You will see that this way you will be forced to instantiate azure clients with `azEnv` inside of the methods (which makes testign harder).